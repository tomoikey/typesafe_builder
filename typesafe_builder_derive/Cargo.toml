[package]
name = "typesafe_builder_derive"
version = "1.3.0"
edition = "2024"
authors = ["tomoikey"]
readme = "README.md"
repository = "https://github.com/tomoikey/typesafe_builder"
categories = ["accessibility", "development-tools", "rust-patterns"]
license = "MIT"
description = "A procedural macro to generate type-safe builder patterns for Rust structs"

[lib]
proc-macro = true

[dependencies]
darling = "0.20.11"
proc-macro2 = "1.0.95"
syn = { version = "2.0.101", features = ["derive", "parsing", "proc-macro"] }
quote = "1.0.40"

[dev-dependencies]
trybuild = "1.0.105"
