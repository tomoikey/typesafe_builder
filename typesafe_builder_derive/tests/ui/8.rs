use typesafe_builder_derive::Builder;

struct _TypesafeBuilderEmpty;
struct _TypesafeBuilderFilled;

fn main() {
    #[derive(Builder)]
    struct Config {
        #[builder(optional)]
        enable_feature: Option<bool>,
        #[builder(optional_if = "enable_feature")]
        feature_config: Option<String>,
    }

    // compile error because feature_config is required when enable_feature is not set (None/false)
    let config = ConfigBuilder::new().build();
} 