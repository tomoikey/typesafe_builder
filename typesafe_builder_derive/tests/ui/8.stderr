error[E0599]: no method named `build` found for struct `ConfigBuilder<_TypesafeBuilderEmpty, _TypesafeBuilderEmpty>` in the current scope
  --> tests/ui/8.rs:16:39
   |
7  |     #[derive(Builder)]
   |              ------- method `build` not found for this struct
...
16 |     let config = ConfigBuilder::new().build();
   |                                       ^^^^^ method not found in `ConfigBuilder<_TypesafeBuilderEmpty, _TypesafeBuilderEmpty>`
   |
   = note: the method was found for
           - `ConfigBuilder<_TypesafeBuilderEmpty, _TypesafeBuilderFilled>`
           - `ConfigBuilder<_TypesafeBuilderFilled, _TypesafeBuilderEmpty>`
           - `ConfigBuilder<_TypesafeBuilderFilled, _TypesafeBuilderFilled>`
